File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/repositories/RepositoriesDependOnDataSourceInterfaces.kt
package com.perrystreet.woof.konsist.repositories

import com.perrystreet.woof.konsist.Assertions.assertTrue
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.KonsistUtils.constructorParameterTypeNames
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class RepositoriesDependOnDataSourceInterfaces : BehaviorSpec() {
    init {
        Given("A Repository") {
            val repositories = KonsistUtils.repositories

            Then("It depends only on data source interfaces and mappers") {
                repositories.assertTrue(message = Message) { repository ->
                    repository.constructorParameterTypeNames.all { type ->
                        (type.startsWith("I") && type.endsWith("DataSource")) || type.endsWith("Mapper")
                    }
                }
            }
        }
    }

    private companion object {
        private val Message = LintRuleMessage(
            rule = "Repositories depend on I*DataSource interfaces and mappers, never on concrete data sources or other repositories.",
            why = """
                Programming against the interface is what lets tests swap the data source for a fake without
                mocks, and keeps repositories from reaching into each other's state.
            """.trimIndent(),
            howToFix = "Inject the I*DataSource interface. Compose repositories in a UseCase instead of injecting one into another.",
            badExample = "class DogsRepository(private val dataSource: DogsLocalDataSource, private val moderation: ModerationRepository)",
            goodExample = "class DogsRepository(private val dataSource: IDogsDataSource, private val pageMapper: DogsPageDTOToDomainMapper)",
        )
    }
}

File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/Repositories/RepositoriesDependOnDataSourceInterfaces.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class RepositoriesDependOnDataSourceInterfaces: QuickSpec {
    override class func spec() {
        Given("A Repository") {
            let repositories = WoofHarmonize.repositories

            Then("It depends only on data source protocols and mappers") {
                repositories.assertTrue(rule: rule) { repository in
                    repository.variables
                        .filter { $0.isStored && $0.isConstant && $0.initializerClause == nil }
                        .allSatisfy { variable in
                            let type = variable.typeAnnotation?.name ?? ""
                            return type.hasSuffix("DataSourceImplementing") || type.hasSuffix("Mapper")
                        }
                }
            }
        }
    }

    private static let rule = Rule(
        description: "Repositories depend on *DataSourceImplementing protocols and mappers, never on concrete data sources or other repositories.",
        rationale: """
            Programming against the protocol is what lets tests swap the data source for a fake without
            mocks, and keeps repositories from reaching into each other's state.
            """,
        fixHint: "Inject the *DataSourceImplementing protocol. Compose repositories in a UseCase instead of injecting one into another.",
        badExample: "final class DogsRepository { private let dataSource: DogsLocalDataSource; private let moderation: ModerationRepository }",
        goodExample: "final class DogsRepository { private let dataSource: DogsDataSourceImplementing; private let pageMapper: DogsPageDTOToDomainMapper }"
    )
}
